server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java

```diff
@@ -343,4 +358,63 @@ public void collect(GarbageCollectionEnvironment gce)
       deleteConfirmed(gce, candidateMap);
     }
   }
+
+  /**
+   * Track metadata rows read to help validate that gc scan has complete information to make a
+   * decision on deleting files
+   */
+  private static class MetadataReadTracker {
+    private boolean hasDir = false;
+    private boolean hasPrevRow = false;
+    private Text row;
+
+    private boolean closed = false;
+
+    public MetadataReadTracker() {
+      this.row = null;
+    }
+
+    private void checkRow() {
+      Preconditions.checkState(hasDir && hasPrevRow,
+          "May not have fully read metadata for row, aborting this run. Validation results: %s",
+          this);
+    }
+
+    public void sawRow(final Text candidate) {
+      Preconditions.checkState(!closed);
+      Objects.requireNonNull(candidate);
+      if (row == null) {
+        row = candidate;
+      } else if (!row.equals(candidate)) {
+        checkRow();
+        hasPrevRow = false;
+        hasDir = false;
+        row = candidate;
+      }
+    }
+
+    public void sawDir() {
+      Preconditions.checkState(!closed);
+      hasDir = true;
+    }
+
+    public void sawPrevRow() {
+      Preconditions.checkState(!closed);
+      hasPrevRow = true;
+    }
+
+    public void sawLastRow() {
+      Preconditions.checkState(!closed);
+      if (row != null) {
+        checkRow();
+      }
+      closed = true;
+    }
+
+    @Override
+    public String toString() {
+      return "MetadataReadCheck{row=" + row + ", hasDir=" + hasDir + ", hasPrevRow=" + hasPrevRow
+          + '}';
+    }
```

Comment:
The name here no longer matches the name of the class. Guava has a nice "ToStringHelper" class. I'm not sure if it's worth adding it here. Adding in more Guava might just be a pain later.

```suggestion
    public String toString() {
      return getClass().getSimpleName() + "{row=" + row + ", hasDir=" + hasDir + ", hasPrevRow=" + hasPrevRow
          + '}';
    }
```
(Note: formatting is almost certainly off in this suggestion.

Reply:
Fixed in bb4f078d46